core/state.py

```diff
@@ -2309,6 +2309,59 @@ def PopContextStack(self):
         return self.ctx_stack.pop()
 
 
+class Procs:
+
+    def __init__(self, mem):
+        # type: (Mem) -> None
+        self.mem = mem
+        self.procs = {}  # type: Dict[str, value.Proc]
+
+    def SetProc(self, name, proc):
+        # type: (str, value.Proc) -> None
+        self.mem.var_stack[0][name] = Cell(False, False, False, proc)
+
+    def SetShFunc(self, name, proc):
+        # type: (str, value.Proc) -> None
+        self.procs[name] = proc
+
+    def Get(self, name):
+        # type: (str) -> value.Proc
```

Comment:
Maybe clarify in the docstring that procs are found before shell functions, since that can matter to the user